tests/bsim/bluetooth/host/l2cap/einprogress/tester/src/tester.c

```diff
@@ -0,0 +1,675 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/sys/__assert.h>
+
+#include <zephyr/net/buf.h>
+#include <zephyr/bluetooth/buf.h>
+
+#include <zephyr/bluetooth/bluetooth.h>
+#include <zephyr/bluetooth/hci.h>
+#include <zephyr/bluetooth/hci_raw.h>
+#include <zephyr/bluetooth/gap.h>
+
+#include "common/bt_str.h"
+
+#include "host/conn_internal.h"
+#include "host/l2cap_internal.h"
+
+#include "babblekit/flags.h"
+#include "babblekit/device.h"
+#include "babblekit/testcase.h"
+
+/* local includes */
+#include "data.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(tester, CONFIG_APP_LOG_LEVEL);
+
+DEFINE_FLAG(is_connected);
+DEFINE_FLAG(flag_l2cap_connected);
+
+static K_FIFO_DEFINE(rx_queue);
+
+#define CMD_BUF_SIZE MAX(BT_BUF_EVT_RX_SIZE, BT_BUF_CMD_TX_SIZE)
+NET_BUF_POOL_FIXED_DEFINE(hci_cmd_pool, CONFIG_BT_BUF_CMD_TX_COUNT, CMD_BUF_SIZE, 8, NULL);
+
+static K_SEM_DEFINE(cmd_sem, 1, 1);
+static struct k_sem acl_pkts;
+static struct k_sem tx_credits;
+static uint16_t peer_mps;
+static uint16_t conn_handle;
+
+static uint16_t active_opcode = 0xFFFF;
+static struct net_buf *cmd_rsp;
+
+struct net_buf *bt_hci_cmd_create(uint16_t opcode, uint8_t param_len)
+{
+	struct bt_hci_cmd_hdr *hdr;
+	struct net_buf *buf;
+
+	LOG_DBG("opcode 0x%04x param_len %u", opcode, param_len);
+
+	buf = net_buf_alloc(&hci_cmd_pool, K_FOREVER);
+	__ASSERT_NO_MSG(buf);
+
+	LOG_DBG("buf %p", buf);
+
+	net_buf_reserve(buf, BT_BUF_RESERVE);
+
+	bt_buf_set_type(buf, BT_BUF_CMD);
+
+	hdr = net_buf_add(buf, sizeof(*hdr));
+	hdr->opcode = sys_cpu_to_le16(opcode);
+	hdr->param_len = param_len;
+
+	return buf;
+}
+
+static void handle_cmd_complete(struct net_buf *buf)
+{
+	struct bt_hci_evt_hdr *hdr;
+	uint8_t status, ncmd;
+	uint16_t opcode;
+
+	struct net_buf_simple_state state;
+
+	net_buf_simple_save(&buf->b, &state);
+
+	hdr = net_buf_pull_mem(buf, sizeof(*hdr));
+
+	if (hdr->evt == BT_HCI_EVT_CMD_COMPLETE) {
+		struct bt_hci_evt_cmd_complete *evt;
+
+		evt = net_buf_pull_mem(buf, sizeof(*evt));
+		status = 0;
+		ncmd = evt->ncmd;
+		opcode = sys_le16_to_cpu(evt->opcode);
+
+	} else if (hdr->evt == BT_HCI_EVT_CMD_STATUS) {
+		struct bt_hci_evt_cmd_status *evt;
+
+		evt = net_buf_pull_mem(buf, sizeof(*evt));
+		status = buf->data[0];
+		ncmd = evt->ncmd;
+		opcode = sys_le16_to_cpu(evt->opcode);
+
+	} else {
+		__ASSERT_NO_MSG(0);
+	}
+
+	LOG_DBG("opcode 0x%04x status %x", opcode, status);
+
+	__ASSERT(status == 0x00, "cmd status: %x", status);
+
+	__ASSERT(active_opcode == opcode, "unexpected opcode %x != %x", active_opcode, opcode);
+
+	if (active_opcode) {
+		active_opcode = 0xFFFF;
+		cmd_rsp = net_buf_ref(buf);
+		net_buf_simple_restore(&buf->b, &state);
+	}
+
+	if (ncmd) {
+		k_sem_give(&cmd_sem);
+	}
+}
```

Comment:
Why are we handling this and more in this file? We can't test everything just via the L2CAP API?

Reply:
![image](https://github.com/user-attachments/assets/9a73e29c-21aa-4c61-bd66-67d64c090135)

Reply:
![image](https://github.com/user-attachments/assets/3d26eb86-eb68-4978-9016-0a1e218e89c4)